File: 2024/Rust/src/day11.rs
use fxhash::FxHashMap as HashMap;

type Parsed = HashMap<u128, u64>;
type Solution = u64;

const PART1_MAX: usize = 25;
const PART2_MAX: usize = 75;

pub fn parse(input: &str) -> Parsed {
    let mut map = HashMap::default();
    for num in input.trim().split_ascii_whitespace() {
        *map.entry(num.parse().unwrap()).or_default() += 1;
    }
    map
}

fn split_middle(num: u128) -> Option<(u128, u128)> {
    // Count number of digits
    let mut temp = num;
    let mut digit_count = 0;
    while temp > 0 {
        digit_count += 1;
        temp /= 10;
    }
    
    if digit_count % 2 != 0 {
        return None;
    }
    
    let divisor = 10_u128.pow(digit_count / 2 );

    let second = num % divisor;
    let first = num / divisor;
    
    Some((first, second))
}

fn blink(last: &mut Parsed, next: &mut Parsed){
    for (key, v) in last.drain() {
        if key == 0 {
            // rule 1
            *next.entry(1).or_default() += v;
        } else if let Some((first, second)) = split_middle(key) {
            // rule 2
            *next.entry(first).or_default() += v;
            *next.entry(second).or_default() += v;
        } else {
            // rule 3
            let key = key * 2024;
            *next.entry(key).or_default() += v;
        }
    }
}

pub fn part1(stones: &Parsed) -> Solution {
    let mut stones = stones.clone();
    let mut next = HashMap::default();
    for _ in 0..PART1_MAX {
        blink(&mut stones, &mut next);
        std::mem::swap(&mut stones, &mut next);
    }

    stones.values().sum()
}

pub fn part2(stones: &Parsed) -> Solution {
    let mut stones = stones.clone();
    let mut next = HashMap::default();
    for _ in 0..PART2_MAX {
        blink(&mut stones, &mut next);
        std::mem::swap(&mut stones, &mut next);
    }

    stones.values().sum()
}

pub fn run(input: &str) {
    let parsed = parse(input);
    println!("Part 1: {}", part1(&parsed));
    println!("Part 2: {}", part2(&parsed));
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str = "125 17";

    const P1_SOLUTION: Solution = 55312;

    #[test]
    fn test_part1() {
        assert_eq!(part1(&parse(INPUT)), P1_SOLUTION);
    } 

    // There is no p2 test :(

    #[test]
    fn test_split_middle() {
        assert_eq!(split_middle(1234), Some((12, 34)));
        assert_eq!(split_middle(12345), None);
        assert_eq!(split_middle(1000), Some((10, 0)));
    }
}

File: 2024/Haskell/src/Day11.hs
{-# LANGUAGE OverloadedStrings #-}

module Day11 (
  day11,
)
where

import AoC
import AoC.Common
import Data.Text (Text)
import Data.Text qualified as Text
import Data.Text.Read (decimal)
import Data.HashMap.Strict qualified as HashMap
import Data.HashMap.Strict (HashMap)
import Debug.Trace (traceShow)

type Parsed = HashMap Integer Int
type Solution = Int

splitDigits :: Integer -> [Integer]
splitDigits 0 = [0]
splitDigits x = reverse $ go x
  where
    go 0 = []
    go n = let (q, r) = n `divMod` 10 in
      r : go q

joinDigits :: [Integer] -> Integer
joinDigits = foldl (\acc d -> acc * 10 + d) 0



-- fn split_middle(num: u128) -> Option<(u128, u128)> {
    -- // Count number of digits
    -- let mut temp = num;
    -- let mut digit_count = 0;
    -- while temp > 0 {
        -- digit_count += 1;
        -- temp /= 10;
    -- }
    -- 
    -- if digit_count % 2 != 0 {
        -- return None;
    -- }
    -- 
    -- let divisor = 10_u128.pow(digit_count / 2 );
-- 
    -- let second = num % divisor;
    -- let first = num / divisor;
    -- 
    -- Some((first, second))
-- }

splitMiddle :: Integer -> Maybe (Integer, Integer)
splitMiddle n =
  let digits = splitDigits (fromIntegral n)
      len = length digits
   in if odd len
        then Nothing
        else
          let (firstDigits, secondDigits) = splitAt (len `div` 2) digits
              first = joinDigits firstDigits
              second = joinDigits secondDigits
           in Just (first, second)

processStones :: HashMap Integer Int -> HashMap Integer Int
processStones stones = HashMap.foldlWithKey' processStone HashMap.empty stones 
  where
    processStone acc stone count =
      case stone of 
        0 -> HashMap.insertWith (+) 0 (count * 2) acc
        _ -> case splitMiddle stone of
          Nothing -> HashMap.insertWith (+) stone count acc
          Just (first, second) ->
            let acc1 = HashMap.insertWith (+) first count acc
                acc2 = HashMap.insertWith (+) second count acc1
            in acc2
  

f2 = foldl' (\acc n -> HashMap.insertWith (+) n 1 acc) HashMap.empty

day11 :: Runner Parsed Solution
day11 =
  let year = 2024
      day = 11
      parser :: Text -> Maybe Parsed
      parser = fmap f2 . traverse (ok . fmap fst . decimal) . Text.splitOn " "
      part1 :: Parsed -> Maybe Solution
      part1 input = return . length $ (!! 25) $ iterate processStones input
      part2 :: Parsed -> Maybe Solution
      part2 input = return . length $ (!! 75) $ iterate processStones input
   in Runner{..}
